Steven/dollar auctions

stevenwal opened the pull request:
This is to allow ThetaVault to sell call options in USDC through Gnosis Auction.

Changes include:
* Added UniswapV3 libraries and interfaces
* Added ThetaVault function to settle Gnosis auction and swap through UniswapV3
* Modification to test configs pointing to new Gamma Oracle for base ThetaVaults
* Modification to base ThetaVault test

kenchangh commented on contracts/interfaces/ISwapRouter.sol:
```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity >=0.7.5;
```
We're using =0.8.4 here

stevenwal replied:
Fixed

kenchangh commented on contracts/libraries/GnosisAuction.sol:
```diff
@@ -118,6 +118,15 @@ library GnosisAuction {
+        internal
+        returns (bytes32 clearingOrder)
+    {
+        clearingOrder = IGnosisAuction(gnosisEasyAuction).settleAuction(
+            auctionID
+        );
```
Since this function is just a single line, could we just call it directly from the contract instead of moving it down to the library?

stevenwal replied:
Removed this from GnosisAuction library.
Directly calling the contract from Vaultlifecycle.

kenchangh commented on contracts/libraries/UniswapRouter.sol:
```diff
@@ -0,0 +1,129 @@
+            ISwapRouter.ExactInputParams({
+                recipient: swapParams.recipient,
+                path: swapParams.path,
+                deadline: block.timestamp + 15,
```
Are we sure 15 seconds is enough? We need to extend this to 10 mins so it does not revert due to the deadline

stevenwal replied:
Fixed. Changed to 10 minutes.

kenchangh commented on contracts/vaults/BaseVaults/RibbonThetaVault.sol:
```diff
@@ -111,57 +143,65 @@ contract RibbonThetaVault is RibbonVault, RibbonThetaVaultStorage {
+        auctionDuration = params._auctionDuration;
+
+        usdcAuction = params._usdcAuction;
+        if (usdcAuction) {
```
This should just be `if (params._usdcAuction)` because reading `usdcAuction` costs more gas (reading from storage)

stevenwal replied:
Fixed

kenchangh commented on contracts/vaults/BaseVaults/RibbonThetaVault.sol:
```diff
@@ -237,6 +277,18 @@ contract RibbonThetaVault is RibbonVault, RibbonThetaVaultStorage {
+    function setPath(bytes memory newPath) external onlyOwner nonReentrant {
+        require(
+            UniswapRouter.checkPath(newPath, USDC, vaultParams.asset),
+            "!newPath"
```
Error message could be improved with `newPath invalid`

stevenwal replied:
Fixed. Changed message to `Invalid swap path`.

kenchangh commented on contracts/libraries/UniswapRouter.sol:
```diff
@@ -0,0 +1,129 @@
+        uint256 amountOut
+    );
+
+    function decodePath(bytes memory path)
```
We can use `bytes calldata path` which would [save some gas](https://ethereum.stackexchange.com/questions/74442/when-should-i-use-calldata-and-when-should-i-use-memory)

stevenwal replied:
For decodePath, we are modifying the variable during the while loop (using your suggestion as reference, but the old code also modifies the path).

```
while (path.hasMultiplePools()) {
  (, tokenOut, ) = path.decodeFirstPool();
  path = path.skipToken();
}
```

stevenwal replied:
Combined `decodePath` and `checkPath` as a single function.

kenchangh commented on contracts/libraries/UniswapRouter.sol:
```diff
@@ -0,0 +1,129 @@
+        // return (address(0), address(0));
+    }
+
+    function getTokenOut(bytes memory path)
```
Same here, can use calldata https://github.com/ribbon-finance/ribbon-v2/pull/167/files#r756666724

stevenwal replied:
This function is redundant, will remove.

stevenwal replied:
Addressed [here](https://github.com/ribbon-finance/ribbon-v2/pull/167#discussion_r757028909).

kenchangh commented on contracts/libraries/UniswapRouter.sol:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    function checkPath(
+        bytes memory path,
```
Same here, can use [calldata](https://github.com/ribbon-finance/ribbon-v2/pull/167/files#r756666724)

stevenwal replied:
Addressed [here](https://github.com/ribbon-finance/ribbon-v2/pull/167#discussion_r757028909).

kenchangh commented on contracts/libraries/UniswapRouter.sol:
```diff
@@ -0,0 +1,129 @@
+        return tokenIn == _tokenIn && tokenOut == _tokenOut;
+    }
+
+    function swap(SwapParams memory swapParams)
```
swapParams can be calldata as well. If we want to optimize it even further, split out the `path` so that it does not involve memory at all. E.g.

```
function swap(SwapParams calldata swapParams, bytes calldata path)
```

stevenwal replied:
Fixed. 
Additionally, no longer use a struct for the params. Some params in the struct were redundant.
```
function swap(
        address recipient,
        address tokenIn,
        uint256 amountIn,
        uint256 minAmountOut,
        address router,
        bytes calldata swapPath
    )
```

kenchangh commented on contracts/libraries/UniswapRouter.sol:
```diff
@@ -0,0 +1,129 @@
+
+        amountOut = ISwapRouter(swapParams.router).exactInput(params);
+
+        emit Swap(
+            swapParams.tokenIn,
+            swapParams.tokenOut,
+            swapParams.amountIn,
+            amountOut
+        );
```
I don't think we need to emit this event because it is already emitted by the  underlying`router.exactInput` call

stevenwal replied:
Fixed. Removed emit event.

kenchangh commented on contracts/libraries/UniswapRouter.sol:
```diff
@@ -0,0 +1,129 @@
+        (tokenIn, tokenOut, fee) = path.decodeFirstPool();
+        bool hasMultiplePools = path.hasMultiplePools();
+
+        if (hasMultiplePools) {
+            path = path.skipToken();
+            while (true) {
+                (, tokenOut, ) = path.decodeFirstPool();
+
+                hasMultiplePools = path.hasMultiplePools();
+
+                if (hasMultiplePools) {
+                    path = path.skipToken();
+                } else {
+                    break;
+                }
+            }
+        }
```
Could we not simplify this logic a bit? Here is my attempt:

```solidity
while (path.hasMultiplePools()) {
  (, tokenOut, ) = path.decodeFirstPool();
  path = path.skipToken();
}
```

stevenwal replied:
Fixed. Works as per your suggestion with slight modification.

```
(address tokenIn, address tokenOut, ) = swapPath.decodeFirstPool();

while (swapPath.hasMultiplePools()) {
    swapPath = swapPath.skipToken();
    (, tokenOut, ) = swapPath.decodeFirstPool();
}
```

kenchangh commented on contracts/libraries/VaultLifecycle.sol:
```diff
@@ -671,6 +672,36 @@ library VaultLifecycle {
+    function settleAuctionAndSwap(
+        address gnosisEasyAuction,
+        uint256 auctionID,
+        bytes memory _path,
```
bytes calldata _path

stevenwal replied:
Fixed.

kenchangh commented on contracts/vaults/BaseVaults/RibbonThetaVault.sol:
```diff
@@ -111,57 +143,65 @@ contract RibbonThetaVault is RibbonVault, RibbonThetaVaultStorage {
+                UniswapRouter.checkPath(params._path, USDC, vaultParams.asset),
+                "!_path"
+            );
+            // require(UniswapRouter.getTokenOut(params._path) == vaultParams.asset, "!path");
+            // (address tokenIn, address tokenOut) = UniswapRouter.decodePath(params._path);
+            // require(tokenIn == address(0), "!tokenIn");
+            // require(tokenOut == address(0), "!tokenOut");
+            // (address tokenIn, address tokenOut) = UniswapRouter.decodePath(params._path);
+            // require(tokenIn == USDC, "!tokenIn");
+            // require(tokenOut == vaultParams.asset, "!tokenOut");
```
Can remove these right?

stevenwal replied:
Removed.

stevenwal commented on contracts/storage/RibbonThetaVaultStorage.sol:
```diff
@@ -25,13 +25,23 @@ abstract contract RibbonThetaVaultStorageV2 {
+    // Auction will be denominated in USDC if true
+    bool public usdcAuction;
+    // Uniswap Router V3 contract for swaps
+    address public uniswapRouter;
```
This wouldn't affect the upgradability of our existing ThetaVault?

kenchangh replied:
It would not, because we are just deploying a new implementation contract that has the immutable variable

stevenwal replied:
Fixed. Added router as immutable.

kenchangh commented on contracts/libraries/UniswapRouter.sol:
```diff
@@ -0,0 +1,55 @@
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+
+    function checkPath(
```
Missing the function documentation comment

stevenwal replied:
Added documentation

kenchangh commented on contracts/libraries/UniswapRouter.sol:
```diff
@@ -0,0 +1,55 @@
+        return tokenIn == validTokenIn && tokenOut == validTokenOut;
+    }
+
+    function swap(
```
Missing the function documentation comment

stevenwal replied:
Added documentation

kenchangh commented on contracts/vaults/BaseVaults/RibbonThetaVault.sol:
```diff
@@ -409,4 +461,26 @@ contract RibbonThetaVault is RibbonVault, RibbonThetaVaultStorage {
+        onlyKeeper
+        nonReentrant
+    {
+        require(isUsdcAuction, "!isUsdcAuction");
+        require(minAmountOut > 0, "!_minAmountOut");
```
Need to add test cases for these requires

stevenwal replied:
Added documentation

kenchangh commented on contracts/vaults/BaseVaults/base/RibbonVault.sol:
```diff
@@ -168,6 +174,7 @@ contract RibbonVault is
         GAMMA_CONTROLLER = _gammaController;
         MARGIN_POOL = _marginPool;
         GNOSIS_EASY_AUCTION = _gnosisEasyAuction;
+        UNISWAP_ROUTER = _uniswapRouter;
```
require(_uniswapRouter != address(0), "_uniswapRouter");

stevenwal replied:
Added this require.

stevenwal commented:
Made a couple changes:
* added new test cases for `settleAuctionAndSwap` and `setSwapPath` functions
* added UNISWAP_FACTORY as immutables, this is to check if pool exist when validating swap path
* make`checkPath` function as public to prevent RibbonThetaVault contract size from exceeding limit
* merge the latest changes from recent PRs

kenchangh commented on test/RibbonThetaVault.ts:
```diff
@@ -428,7 +551,7 @@ function behavesLikeRibbonOptionsVault(params: {
       firstOptionExpiry = moment(latestTimestamp * 1000)
         .startOf("isoWeek")
-        .add(chainId === CHAINID.ETH_MAINNET ? 1 : 0, "week")
+        // .add(chainId === CHAINID.ETH_MAINNET ? 1 : 0, "week")
```
Why did we comment this out?

stevenwal replied:
I'm getting the wrong otoken address if I move it one week forward. I might be wrong here, but this might be because of the new blocknumber, so the next option is on the same week, not the following week.

kenchangh replied:
We can remove it because it's not used

kenchangh commented on contracts/vaults/BaseVaults/RibbonThetaVault.sol:
```diff
@@ -237,6 +269,19 @@ contract RibbonThetaVault is RibbonVault, RibbonThetaVaultStorage {
+        onlyOwner
+        nonReentrant
+    {
+        require(_checkPath(newSwapPath), "Invalid swap path");
```
+ `require(isUsdcAuction)`

kenchangh commented on contracts/vaults/BaseVaults/RibbonThetaVault.sol:
```diff
@@ -387,8 +432,8 @@ contract RibbonThetaVault is RibbonVault, RibbonThetaVaultStorage {
         auctionDetails.gnosisEasyAuction = GNOSIS_EASY_AUCTION;
-        auctionDetails.asset = vaultParams.asset;
-        auctionDetails.assetDecimals = vaultParams.decimals;
+        auctionDetails.asset = isUsdcAuction ? USDC : vaultParams.asset;
```
Since `isUsdcAuction` reads from storage and is used twice, we can save gas by caching it in a variable first

```solidity
bool _isUsdcAuction = isUsdcAuction;
auctionDetails.asset = _isUsdcAuction ? USDC : vaultParams.asset;
```

kenchangh commented on test/RibbonThetaVault.ts:
```diff
@@ -947,6 +1110,118 @@ function behavesLikeRibbonOptionsVault(params: {
+            DEX_FACTORY[chainId]
+          );
+
+          let pool = await dexFactory.getPool(asset, asset, 10000);
+          if (pool == constants.AddressZero) {
```
Isn't `pool` guaranteed to be `address(0)`? Since you are doing `getPool(asset, asset)` which does not exist

stevenwal replied:
Yes, this is to prove that the `dexFactory` indeed returns a zero address. And therefore matches the result we are getting from `setSwapPath`.

kenchangh commented on test/RibbonThetaVault.ts:
```diff
@@ -3136,6 +3492,101 @@ function behavesLikeRibbonOptionsVault(params: {
+          await expect(
+            vault.connect(keeperSigner).settleAuctionAndSwap("0")
+          ).to.be.revertedWith("!minAmountOut");
+        });
```
Nit new line between it block

kenchangh commented on test/RibbonThetaVault.ts:
```diff
@@ -3192,3 +3643,23 @@ async function depositIntoVault(
   }
 }
+
+function encodePath(tokenAddresses, fees) {
```
Let's move this out to a test/helpers/utils.ts

kenchangh commented:
We need to ignore checking in `contracts/.DS_Store`

kenchangh commented on test/RibbonThetaVault.ts:
```diff
@@ -132,18 +146,88 @@ describe("RibbonThetaVault", () => {
     minimumSupply: BigNumber.from("10").pow("3").toString(),
     expectedMintAmount: BigNumber.from(
-      chainId === CHAINID.ETH_MAINNET ? "5263157894" : "2702702702"
+      "2702702702" //chainId === CHAINID.ETH_MAINNET ? "5263157894" : "2702702702"
```
Why are we commenting this line out? This is for checking avax

kenchangh commented on test/RibbonThetaVault.ts:
```diff
@@ -466,7 +590,8 @@ function behavesLikeRibbonOptionsVault(params: {
       secondOptionExpiry = moment(latestTimestamp * 1000)
         .startOf("isoWeek")
-        .add(chainId === CHAINID.ETH_MAINNET ? 2 : 1, "week")
+        // .add(chainId === CHAINID.ETH_MAINNET ? 2 : 1, "week")
```
We can remove this if it's commented out

kenchangh reviewed (approved):
LGTM 🚢
